8373203: Genshen: Non-strong reference leak in old gen

earthling-amzn opened the pull request:
The generational mode for Shenandoah will collect _referents_ for the generation being collected. For example, if we have a young reference pointing to an old referent, that young reference will be processed after we finish marking the old generation. This presents a problem for discovery.
  
When the young mark _encounters_ a young reference with an old referent, it cannot _discover_ it because old marking hasn't finished. However, if it does not discover it, the old referent will be strongly marked. This, in turn, will prevent the old generation from clearing the referent (if it even reaches it again during old marking).
  
To solve this, we let young reference processing discover the old reference by having it use the old generation reference processor to do so. This means the old reference processor can have a discovered list that contains young weak references. If any of these young references reside in a region that is collected, old reference processing will crash when it processes such a reference. Therefore, we add a method `heal_discovered_lists` to traverse the discovered lists after young evacuation is complete. The method will replace any forwarded entries in the discovered list with the forwardee.

This PR also extends whitebox testing support for Shenandoah, giving us the ability to trigger young/old collections and interrogate some properties of heaps and regions.

- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[ \] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8373203](https://bugs.openjdk.org/browse/JDK-8373203): Genshen: Non-strong reference leak in old gen (**Bug** - P4)


### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - **Reviewer**) 🔄 Re-review required (review applies to [f621b70c](https://git.openjdk.org/jdk/pull/28810/files/f621b70c045234d943585252fb97d0d390ac19a3))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28810/head:pull/28810` \
`$ git checkout pull/28810`

Update a local copy of the PR: \
`$ git checkout pull/28810` \
`$ git pull https://git.openjdk.org/jdk.git pull/28810/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28810`

View PR using the GUI difftool: \
`$ git pr show -t 28810`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28810.diff">https://git.openjdk.org/jdk/pull/28810.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28810#issuecomment-3648633084)
</details>

earthling-amzn committed:
Add test for genshen reference processing

Also, enhance whitebox support for Shenandoah

earthling-amzn committed:
Clean up test, increase likelihood of cross generational references

earthling-amzn committed:
WIP: checkpoint

earthling-amzn committed:
Add comments, fix typo, survive multiple young GCs

earthling-amzn committed:
Add missing newline

earthling-amzn committed:
Oops, initialize new field

earthling-amzn committed:
Oops, change name of class in test xdoc

earthling-amzn committed:
Use Events::log, not Event for simple log messages

earthling-amzn committed:
Consolidate management of bootstrap cycle configuration

earthling-amzn committed:
Do not bypass card barrier when healing discovered list

earthling-amzn committed:
Clear bootstrap mode for full GC that might have bypassed degenerated cycle

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Heal discovered oops in common place for degen and concurrent update refs

earthling-amzn committed:
Factor duplicate code into shared method

earthling-amzn committed:
Fix comment

earthling-amzn committed:
Heal old discovered lists in parallel

earthling-amzn committed:
Sort includes

stefank reviewed (commented):
I added a few style proposals to make the Shenandoah sections look more consistent with the surrounding style in the shared code.

earthling-amzn committed:
Fix idiosyncratic white space in whitebox

Co-authored-by: Stefan Karlsson <stefan.karlsson@oracle.com>

kdnilsen reviewed (approved):
Thanks for chasing this problem down.  Very important fix.

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Configure thread local mark closure on delegated old reference processor

earthling-amzn committed:
Heal discovered lists for any young collection coincides with old marking

earthling-amzn commented:
This change has now passed internal testing pipelines several times.

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn commented:
Consider a situation where an old and young reference both point to an old referent:
```
young_reference       old_reference
              \           /
               \         /
               old_referent
```
Genshen "bootstraps" old marking with a young collection. During young generation marking, oop iteration will "encounter" `young_reference`, and ask our reference processor to "discover" it. If the reference processor _discovers_ `old_referent`, then the reference processor is responsible for it and the mark thread will _not_ mark through the referent. Conversely, if the reference processor does _not_ discover the referent, then oop iteration will mark the referent and trace through it. This behavior is baked pretty deep into oop iteration and any changes in this code would affect all collectors.

The young reference processor cannot _discover_ old referents because it will not have complete mark information when it comes time to process the discovered list. By not discovering the old referent, the mark thread will mark and trace through the old referent. Following the bootstrap, when old marking encounters `old_reference`, it sees that `old_referent` is already strongly marked and so similarly does not _discover_ it. We could continue to wait for `young_reference` to die or be promoted, but you can see that a single `young_reference` to any `old_referent` can prevent the reference from being cleared.

This PR has the young reference processor defer discovery to the old reference processor. This is simple enough, except that the young references now live in the old reference processor's discovered list. This list runs through the heap so these young references can be evacuated. This requires us to update references for the old processor's discovered list. Note that the list already updates the card table when young references are placed on the discovered list, so they become roots for subsequent young collections by virtue of being in the remembered set.

ysramakrishna commented:
> ... unless there is a good practical reason to fix it in this manner.

Got more info from William offline; I am going to take a closer look at the changes in light of that, the explanation he provided above, and the example provided by the submitter. Thanks William!

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Heal discovered lists during evacuation, use full strength LRB

earthling-amzn committed:
Allow young marking access to old mark queues during entirety of old marking

earthling-amzn committed:
Mark references that have been discovered by the old reference processor

This will also make sure they are evacuated, so we should be able to go back to using the `lrb` function defined in shenandoahReferenceProcessor.cpp

earthling-amzn committed:
Mark all fields except referent when keeping references alive for old reference processing

earthling-amzn committed:
Oops, forgot to initialize new member

earthling-amzn committed:
Mark discovered lists for old generation during remembered set scan

earthling-amzn committed:
Fix dependency issues

earthling-amzn committed:
Put templated methods in inline file

earthling-amzn committed:
Remove too strong assert, rename method

earthling-amzn committed:
Instance ref klass doesn't want a reference discoverer configured when we don't iterate referents

earthling-amzn committed:
Fix compilation errors

earthling-amzn committed:
Fix delegate

earthling-amzn committed:
Fix off-by-one error :facepalm:

earthling-amzn committed:
mark the head of the list

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Fix merge error

earthling-amzn committed:
Fix warnings on macos build

earthling-amzn committed:
Add copyright header, test with compressed oops enabled

earthling-amzn committed:
Fix whitebox exception messages

earthling-amzn committed:
Use atomic wrapper for worker id iterator

earthling-amzn committed:
Remove unused/undefined method declaration

earthling-amzn committed:
Fix copy/pasta test summary, add @bug id

earthling-amzn committed:
Put test case xdoc above package statement

earthling-amzn committed:
Test style/correctness improvements

earthling-amzn committed:
Make lrb helper function inline, take it out of the header

earthling-amzn committed:
Remove unnecessary and misleading code

earthling-amzn committed:
Fix idiosyncratic formatting

earthling-amzn committed:
heal the last (self) reference in the discovered list

earthling-amzn committed:
Remove unused field

earthling-amzn committed:
Add in headers pulled in transitively

earthling-amzn committed:
Move inline functions in header out of global namespace

earthling-amzn committed:
Document why we must always mark the young reference

earthling-amzn committed:
Use more suited data structure, better argument handling

earthling-amzn committed:
Use atomic wrapper for index iterator

earthling-amzn committed:
Include guard comment consistency

earthling-amzn committed:
Fix stale comment

earthling-amzn committed:
Fix formatting and capture for heal_discovered_lists lambda

earthling-amzn committed:
Add coverage for compressed oops disabled

earthling-amzn committed:
Avoid redundant LRB calls and recalculations, clearer names for variables

earthling-amzn committed:
Remove unused whitebox method

earthling-amzn committed:
Do not return meaningless boolean

earthling-amzn committed:
Remove extra newline at end of file

earthling-amzn committed:
Remove unused method

earthling-amzn committed:
Avoid apply LRB to discovered list twice

earthling-amzn committed:
Avoid double marking discovered list elements

earthling-amzn committed:
Add missing transitive include

earthling-amzn committed:
Initialize all members in initializer list

earthling-amzn committed:
Remove non-existent test library path

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing

earthling-amzn committed:
Merge remote-tracking branch 'jdk/master' into fix-old-reference-processing
